src/precice/impl/WatchIntegral.cpp

```diff
@@ -0,0 +1,201 @@
+#include "WatchIntegral.hpp"
+#include <Eigen/Core>
+#include "logging/LogMacros.hpp"
+#include "mesh/Data.hpp"
+#include "mesh/Edge.hpp"
+#include "mesh/Mesh.hpp"
+#include "mesh/SharedPointer.hpp"
+#include "mesh/Triangle.hpp"
+#include "mesh/Vertex.hpp"
+#include "utils/MasterSlave.hpp"
+
+namespace precice {
+namespace impl {
+
+WatchIntegral::WatchIntegral(
+    mesh::PtrMesh      meshToWatch,
+    const std::string &exportFilename,
+    bool               isScalingOn)
+    : _mesh(std::move(meshToWatch)),
+      _txtWriter(exportFilename),
+      _isScalingOn(isScalingOn)
+{
+  PRECICE_ASSERT(_mesh);
+
+  if (not utils::MasterSlave::isSlave()) {
+    _txtWriter.addData("Time", io::TXTTableWriter::DOUBLE);
+  }
+
+  for (size_t i = 0; i < _mesh->data().size(); i++) {
+    _dataToExport.push_back(_mesh->data()[i]);
+
+    if (not utils::MasterSlave::isSlave()) {
+      if (_dataToExport[i]->getDimensions() > 1) {
+
+        io::TXTTableWriter::DataType vectorType = _dataToExport[i]->getDimensions() == 2
+                                                      ? io::TXTTableWriter::VECTOR2D
+                                                      : io::TXTTableWriter::VECTOR3D;
+        _txtWriter.addData(_dataToExport[i]->getName(), vectorType);
+      } else {
+        _txtWriter.addData(_dataToExport[i]->getName(), io::TXTTableWriter::DOUBLE);
+      }
+    }
+  }
+
+  if (not utils::MasterSlave::isSlave()) {
+    _txtWriter.addData("SurfaceArea", io::TXTTableWriter::DOUBLE);
+  }
+
+  if ((not _isScalingOn) and (_mesh->edges().empty() or _mesh->triangles().empty())) {
+    PRECICE_WARN("Watch-integral is configured with scaling option on; however, mesh " << _mesh->getName() << " does not contain connectivity information. Therefore, the integral will be calculated without scaling.");
+  }
```

Comment:
```suggestion
  if ((_isScalingOn) and (_mesh->edges().empty() or _mesh->triangles().empty())) {
    PRECICE_WARN("Watch-integral is configured with scaling option on; however, mesh " << _mesh->getName() << " does not contain connectivity information. Therefore, the integral will be calculated without scaling.");
  }
  Another problem I see is that in 2D the set of triangles is always empty, right?
```

Reply:
Right. So basically, I think we can directly say if edge container is empty, there is no connectivity.

Reply:
yes, that solves the second problem. but the `not` is also still wrong, right?

Reply:
Oh, I missed it sorry